src/Serilog.Sinks.Elasticsearch/Sinks/ElasticSearch/ElasticsearchSinkState.cs

```diff
@@ -0,0 +1,156 @@
+        private readonly string _templateName;
+        private readonly string _templateMatchString;
+        private static readonly Regex IndexFormatRegex = new Regex(@"^(.*)(?:\{0\:.+\})(.*)$");
+
+		public ElasticsearchSinkOptions Options { get { return this._options; }}
+		public IElasticsearchClient Client { get { return this._client; }}
+		public ITextFormatter Formatter { get { return this._formatter; }}
+
+
+		private ElasticsearchSinkState(ElasticsearchSinkOptions options)
+		{
+            if (string.IsNullOrWhiteSpace(options.IndexFormat)) throw new ArgumentException("options.IndexFormat");
+            if (string.IsNullOrWhiteSpace(options.TypeName)) throw new ArgumentException("options.TypeName");
+            if (string.IsNullOrWhiteSpace(options.TemplateName)) throw new ArgumentException("options.TemplateName");
+
+            this._templateName = options.TemplateName;
+            this._templateMatchString = IndexFormatRegex.Replace(options.IndexFormat, @"$1*$2");
+            
+            _indexDecider = options.IndexDecider ?? ((@event, offset) => string.Format(options.IndexFormat, offset));
+
+            _typeName = options.TypeName;
+			_options = options;
+
+            var configuration = new ConnectionConfiguration(options.ConnectionPool)
+                .SetTimeout(options.ConnectionTimeout)
+                .SetMaximumAsyncConnections(20);
+
+            if (options.ModifyConnectionSetttings != null)
+                configuration = options.ModifyConnectionSetttings(configuration);
+
+            _client = new ElasticsearchClient(configuration, connection: options.Connection, serializer: options.Serializer);
+            _formatter = options.CustomFormatter ?? new ElasticsearchJsonFormatter(
+                formatProvider: options.FormatProvider,
+                renderMessage: true,
+                closingDelimiter: string.Empty,
+                serializer: options.Serializer,
+                inlineFields: options.InlineFields
+            );
+
+			this._registerTemplateOnStartup = options.AutoRegisterTemplate;
+		}
+
+
+		public string Serialize(object o)
+		{
+			var bytes = _client.Serializer.Serialize(o, SerializationFormatting.None);
+			return Encoding.UTF8.GetString(bytes);
+		}
+
+		public string GetIndexForEvent(LogEvent e, DateTimeOffset offset)
+		{
+			return this._indexDecider(e, offset);
+		}
+
+        /// <summary>
+        /// Register the elasticsearch index template if the provided options mandate it.
+        /// </summary>
+        public void RegisterTemplateIfNeeded()
+        {
+            if (!this._registerTemplateOnStartup) return;
+            var result = this._client.IndicesPutTemplateForAll<VoidResponse>(this._templateName, new
+            {
+                template = this._templateMatchString,
+                settings = new Dictionary<string, string>
+                {
+                    {"index.refresh_interval", "5s"}
```

Comment:
best practice for shards/replica's is the default, we could open it up so folks can set it though. 

compression is on by default since atleast 0.90.